File: sample-code/sorting/MergeSort.java
package com.lits.ads002;

import java.util.Arrays;

public class Main {

    public static void main(String[] args) {
        Integer[] array = { 5, 13, 6, 3, 1, 5, 6, 7, 89, 9, 64 };
        mergeSort(array);
        System.out.println(Arrays.toString(array));
    }

    private static void mergeSort(Comparable[] array) {
        Comparable[] mergeResults = new Comparable[array.length];
        mergeSortRecursive(array, mergeResults, 0, array.length - 1);
    }

    private static void mergeSortRecursive(Comparable[] array, Comparable[] mergeResults, int left, int right) {
        if (left < right) {
            int center = (left + right) / 2;
            mergeSortRecursive(array, mergeResults, left, center);
            mergeSortRecursive(array, mergeResults, center + 1, right);
            merge(array, mergeResults, left, center + 1, right);
        }
    }

    private static void merge(Comparable[] array, Comparable[] mergeResults, int leftBegin, int rightBegin, int rightEnd) {
        int leftEnd = rightBegin - 1;
        int leftReadPos = leftBegin;
        int rightReadPos = rightBegin;
        int resultWritePos = leftBegin;

        while (leftReadPos <= leftEnd && rightReadPos <= rightEnd) {
            if (array[leftReadPos].compareTo(array[rightReadPos]) < 0) {
                mergeResults[resultWritePos++] = array[leftReadPos++];
            } else {
                mergeResults[resultWritePos++] = array[rightReadPos++];
            }
        }

        while (leftReadPos <= leftEnd) {
            mergeResults[resultWritePos++] = array[leftReadPos++];
        }

        while (rightReadPos <= rightEnd) {
            mergeResults[resultWritePos++] = array[rightReadPos++];
        }

        for (int i = leftBegin; i <= rightEnd; i++) {
            array[i] = mergeResults[i];
        }
    }
}

File: sample-code/sorting/sorting_benchmark/merge_sort.py
from sorting_algorithm import SortingAlgorithm


class MergeSortAlgorithm(SortingAlgorithm):
    def sort(self, array):
        merge_results = [0] * len(array)
        self.merge_sort_recursive(array, merge_results, 0, len(array) - 1)

    def merge_sort_recursive(self, array, merge_results, left, right):
        if left < right:
            middle = (left + right) // 2
            self.merge_sort_recursive(array, merge_results, left, middle)
            self.merge_sort_recursive(array, merge_results, middle + 1, right)
            self.merge(array, merge_results, left, middle + 1, right)

    def merge(self, array, merge_results, left_begin, right_begin, right_end):
        left_end = right_begin - 1
        left_read_pos = left_begin
        right_read_pos = right_begin
        result_write_pos = left_begin

        while left_read_pos <= left_end and right_read_pos <= right_end:
            if self.compare(array[left_read_pos], array[right_read_pos]):
                merge_results[result_write_pos] = array[left_read_pos]
                left_read_pos += 1
            else:
                merge_results[result_write_pos] = array[right_read_pos]
                right_read_pos += 1
            result_write_pos += 1

        while left_read_pos <= left_end:
            merge_results[result_write_pos] = array[left_read_pos]
            left_read_pos += 1
            result_write_pos += 1

        while right_read_pos <= right_end:
            merge_results[result_write_pos] = array[right_read_pos]
            right_read_pos += 1
            result_write_pos += 1

        array[left_begin:right_end + 1] = merge_results[left_begin:right_end + 1]
